ocaml/alerts/certificate/certificate_check.ml

```diff
@@ -38,142 +23,73 @@ let get_certificate_attributes rpc session_id =
-  | _, Internal _ ->
-      ( cert
-      , Some (body expiring, Api_messages.host_internal_certificate_expiring_30)
-      )
+let alert_message_cls_and_obj_uuid rpc session_id cert =
+  match cert with
+  | Host (host, _) | Internal (host, _) ->
+      (`Host, XenAPI.Host.get_uuid ~rpc ~session_id ~self:host)
+  | CA (cert, _) ->
+      (`Certificate, XenAPI.Certificate.get_uuid ~rpc ~session_id ~self:cert)
 
-let execute rpc session_id existing_messages (cert, alert) =
-  (* CA-342551: messages need to be deleted if the pending alert regard the
-     same host and has newer, updated information.
-     If the pending alert has the same metadata as the existing host message
-     it must not be emmited and the existing message must be retained, this
-     prevents changing the message UUID.
-     In the case there are alerts regarding the host but no alert is pending
-     they are not destroyed since no alert is automatically dismissed. *)
-  match alert with
-  | Some (body, (alert, priority)) -> (
-    try
-      let cls, obj_uuid =
-        match cert with
-        | Host (host, _) | Internal (host, _) ->
-            (`Host, XenAPI.Host.get_uuid ~rpc ~session_id ~self:host)
-        | CA (cert, _) ->
-            ( `Certificate
-            , XenAPI.Certificate.get_uuid ~rpc ~session_id ~self:cert
-            )
-      in
-      let messages_in_host =
-        List.filter
-          (fun (_, record) -> record.API.message_obj_uuid = obj_uuid)
-          existing_messages
+let get_expiry = function
+  | Host (_, valid_till) | Internal (_, valid_till) | CA (_, valid_till) ->
+      Date.of_float (Date.to_float valid_till +. 1.)
+
+let certificates_to_expiry_message_info_list rpc session_id certificates =
+  List.map
+    (fun cert ->
+      let message_cls, message_obj_uuid =
+        alert_message_cls_and_obj_uuid rpc session_id cert
       in
-      let is_outdated (_ref, record) =
-        record.API.message_body <> body
-        || record.API.message_name <> alert
-        || record.API.message_priority <> priority
+      let obj_description = certificate_description cert in
+      let message_sent_on_remaining_days_list =
+        message_sent_on_remaining_days_list cert
       in
-      let outdated, current = List.partition is_outdated messages_in_host in
-
-      List.iter
-        (fun (self, _) -> XenAPI.Message.destroy ~rpc ~session_id ~self)
-        outdated ;
-      if current = [] then
-        let (_ : [> `message] API.Ref.t) =
-          XenAPI.Message.create ~rpc ~session_id ~name:alert ~priority ~cls
-            ~obj_uuid ~body
-        in
-        ()
-    with Api_errors.(Server_error (_handle_invalid, _)) ->
-      (* this happens when the host reference is invalid *)
-      ()
-  )
-  | None ->
-      ()
+      let expiry = get_expiry cert in
+      Expiry_alert.
+        {
+          message_cls
+        ; message_obj_uuid
+        ; obj_description
+        ; message_sent_on_remaining_days_list
+        ; expiry
```

Comment:
I would suggest to replace `obj_description`, `message_sent_on_remaining_days_list` and `expiry` with `msg_name_list` in the interface. The purpose is to minimize/simplify the interface.
Leave `generate_alert` to individual callers (The implementation can be provided as a generic helper function). And the `msg_name_list` would be just a `string list`.

Reply:
> Leave generate_alert to individual callers

That is another option: do not generate `alert` within the lib, while the caller need to generate the `alert` by itself following some required format. And provide the generated alert to the lib in some way(as an parameter or implemented a `generate_alert` interface of the lib functor)

> The implementation can be provided as a generic helper function

To provide a generic helper function to generate alert, I think this requires these parameter again.

Reply:
> To provide a generic helper function to generate alert, I think this requires these parameter again.
No. The helper function will accept much fewer parameters.

Reply:
And the whole thing would be something like:
```
type alert_msg_t = {
    cls: ...
  ; obj_uuid: string
  ; filter_names: string list
  ; body: string
  ; name: string
  ; prio: int64
}

all_messages rpc session_id 
|> List.iter (fun x ->
    all_msgs
    |> List.filter (fun (_ref, record) ->
        record.API.message_obj_uuid = x.obj_uuid
        && List.mem record.API.message_name x.filter_names
    )
    |> List.partition (fun (_ref, record) ->
        record.API.message_body <> x.body
        || record.API.message_name <> x.name
        || record.API.message_priority <> x.prio
    )
    |> (fun (outdated, current) ->
        List.iter
          (fun (self, _) -> XenAPI.Message.destroy ~rpc ~session_id ~self)
          outdated ;
        match current with
        | [] -> (
            XenAPI.Message.create ~rpc ~session_id ~name:x.name ~priority:x.prio ~cls:x.cls
              ~obj_uuid:x.obj_uuid ~body:x.body
            |> ignore
        )
        | _ ->
            ()
    )
)
```

Reply:
I see the above logic is the logic of `update_message_internal`, and I think the `body`, `name`, `prio` are from the caller self generated alert, and this only processed 1 alert.
Then 
1. How to provide a generic helper function for callers to generate alert?
2. To process more than 1 alert at a time, I think a list of `alert_msg_t ` will be needed, I agree the parameters are a little simpler than the current implementation, while callers need to generate the alert themselves. From the callers' point of view, I think the current implementation is a little easier.

Reply:
I think the function `alert` now plays two parts: `generate alert` and `update existing alert messages`.
Combining them in one function would result in verbose input parameters as it has to be a union of two inputs.

Reply:
But I don't have better idea except splitting them which may get the usage of the lib complicated, for example, calling two lib functions instead of only one.